### PR TITLE
feat: Pause other deliveries by the same user when a new session is started

### PR DESCRIPTION
**Pull request by hectoras**
**Associated Jira issue:** [RFE-748](https://oat-sa.atlassian.net/browse/RFE-748)

[RFE-748]: https://oat-sa.atlassian.net/browse/RFE-748?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jsconan — COMMENTED** [review]
The idea of pausing the existing delivery executions is good. However, I'd rather see the mechanism located at the taoQtiTest extension to keep the concerns separated, then have only the trigger in here.

BTW, the PR is targeting `master`. I guess this is a mistake. Meanwhile I created the integration branch: `feat/RFE-748-integration-branch`. I suggest rebasing the PR.

File: controller/Review.php [hunk 1]
```diff
@@ -136,6 +161,110 @@ public function index(): void
         $this->composeView('delegated-view', $data, 'pages/index.tpl', 'tao');
     }
 
+    private function pauseConcurrentSessions(
```
Comment (gabrielfs7) [inline controller/Review.php:164]: This smells like a complete new Service, not a controller method. Wdyt?

File: controller/Review.php [hunk 2]
```diff
@@ -118,6 +138,11 @@ public function index(): void
         } else {
             $execution = $finder->findDeliveryExecution($launchData);
         }
+
+        if ($this->isPausingConcurrentSessionsEnabled()) {
+            $this->pauseConcurrentSessions($execution);
+        }
```
Comment (gabrielfs7) [inline controller/Review.php:144]: I would also move the decision of what to do to the new service (locate on other extension most probably delivery extension).

```suggestion
            $this->pauseService->pauseConcurrentSessions($execution);
```

File: controller/Review.php [hunk 3]
```diff
@@ -136,6 +161,110 @@ public function index(): void
...
+                        '%s: Unable to pause delivery execution %s: %s',
+                        self::class,
+                        $executionId,
+                        $e->getMessage()
+                    )
+                );
+            }
+        }
+
+        $logger->debug(
```
Comment (gabrielfs7) [inline controller/Review.php:227]: We might be abusing a bit of debug in this very PR, lets check if we can remove some of them. This increases logs capacity and costs in several envs QA AWS, GCP, etc, so it is not activated only on local dev. We have to be mindful on adding logs.

File: controller/Review.php (hunk 1 above)
Comment (hectoras) [reply controller/Review.php:164]: I though about that, but eventually found the changes to be too cumbersome at that time and went back to keep the logic in the controller. The problem lies with the later call to store the pause reason as part of the user session, which would bound the service to the user session (and thus making it stateful).

```php
 $this->setSessionAttribute(
            "pauseReason-{$execution->getIdentifier()}",
            PauseService::PAUSE_REASON_CONCURRENT_TEST
        );
```

I'll try again to move it into a service (probably returning the IDs for the paused executions so the controller can still store the pause reason for each test in the user session).

Comment (hectoras) [reply controller/Review.php:171]: This should make it easier to move the code to pause all concurrent sessions into a service: 216481426b3c2b5fd5654344cb09bdd058e0c4e4

**Review by gabrielfs7 — APPROVED** [review]

**Pull request merged**